tests/virt_autotest/virt_autotest_base.pm

```diff
@@ -55,67 +49,27 @@ sub generateXML {
             $fail_nums += 1;
         }
     }
-
     diag '@{$self->{success_guest_list}} content is: ' . Dumper(@{$self->{success_guest_list}});
-
-    my $count = $pass_nums + $fail_nums + $skip_nums;
-    $writer->startTag(
-        'testsuites',
-        error    => "0",
-        failures => "$fail_nums",
-        name     => $self->{product_name},
-        skipped  => "0",
-        tests    => "$count",
-        time     => ""
-    );
-    $writer->startTag(
-        'testsuite',
-        error     => "0",
-        failures  => "$fail_nums",
-        hostname  => "`hostname`",
-        id        => "0",
-        name      => $self->{product_tested_on},
-        package   => $self->{package_name},
-        skipped   => "0",
-        tests     => $case_num,
-        time      => "",
-        timestamp => "2016-02-16T02:50:00"
-    );
-
-    foreach my $item (keys(%my_hash)) {
-        if ($my_hash{$item}->{status} =~ m/PASSED/) {
-            $case_status = "success";
-        }
-        elsif ($my_hash{$item}->{status} =~ m/SKIPPED/ && $item =~ m/iso/) {
-            $case_status = "skipped";
+    ###Load instance attributes into %xmldata
+    my %xmldata;
+    foreach (keys %{$self}) {
+        if (defined($self->{$_})) {
+            if (ref($self->{$_}) eq 'HASH') {
+                %{$xmldata{$_}} = %{$self->{$_}};
+            }
+            elsif (ref($self->{$_}) eq 'ARRAY') {
+                @{$xmldata{$_}} = @{$self->{$_}};
+            }
+            else {
+                $xmldata{$_} = $self->{$_};
+            }
```

Comment:
This xmldata from line 53 to line 69 seems to be dup with virsh_autotest_base.pm line 50-67,  please think about reuse solution via function.

Reply:
Please refer to my above comments, thanks for understanding.

Reply:
I do not think you get my point here. Please cleanup code, it is clearly code duplication which can be handled by reuse a function.

Reply:
foreach loop is already a very compact form to iterate the things you want. So you not need to create a new subroutine to wrap up only "foreach" in it, although foreach might be called more than once. It does not only involves a new subroutine, it is also related to class instance. So, again, "Any modification here requires further tuning or tweak in other subroutines or modules. This is beyond the scope of this PR. In fact, there are many imperfection in the original implementation, so it should be kept as what it was currently in this PR. This is what I mean."

Reply:
> foreach loop is already a very compact form to iterate the things you want. So you not need to create a new subroutine to wrap up only "foreach" in it, although foreach might be called more than once. 

THIS IS THE POINT. If something are called more than once, then it should not be duplicated just everytime. Putting it in a function in a lib file, just as generateXML function does. 

>It does not only involves a new subroutine, it is also related to class instance. 

The above suggestion can solve it already.

>So, again, "Any modification here requires further tuning or tweak in other subroutines or modules. This is beyond the scope of this PR. In fact, there are many imperfection in the original implementation, so it should be kept as what it was currently in this PR. This is what I mean."

I am sorry , but I can not accept it. You said many times many changes are not in this scope? Then what is exactly in this scope? We already accept that some major improvements can be done in a new PR, eg support detailed subtest status result. But for this specific suggestion to remove code duplication, this is exactly in this PR's scope. This has nothing to do with enhancement. It is just this code implementation way. Yes, nothing is perfect. But we should try our best to remove the OBVIOUS imperfection. 
IMO this change should be a small change, wrap these lines into a function in a library file like virt_utils.pm, then use it directly in these two classes virt_autotest_base and virsh_autotest_base.

Reply:
Actually, it does not deserve a new subroutine in lib file IMHO. Because according to subroutine definition:
##################################################################
A subroutine (or routine, function, procedure, macro, etc.) is, at its heart, a named chunk of work. It’s shorthand that allows you to think about your problem in bigger chunks. Bigger chunks means two things:

    You can break the problem up into smaller problems that you can solve independently.
    You can use these solutions to solve your overall problem with greater confidence.
#####################################################################
So you can see the foreach loop itself is already an "atomic problem" itself. If it is wrapped up in another subroutine, it does not improve or alter anything. It is just moved to another place and given a new name. And at the same time, it decreases the "confidence", because it does not reflect the operations that are closely related to class instance anymore. It stands in between, half like a common function, half like a class member, although it might not be called by other modules at all. It becomes useless, so it decreases "confidence".

It should be improved together with upcoming feature implementations, because it is closely related to class instance. According to your saying, yes, "nothing is perfect" and "we try our best", these two do not conflict with each other, and sometimes we need to accept it before evolving to the perfection.

Reply:
I am sorry this can not convince me. A function of over 10 lines deserves to be a function already. Since you said this can be improved together with coming PRs, I can agree with this.

Reply:
But these 10 lines might be useless to all other modules or libs. It should be resolved further within class implementation IMHO. I can classify all these to "junit log enhancement" from what I have seen till now. So any upcoming improvements fall within this category deserve a new or separate PR, which also makes more sense.